crates/transport/src/layers/fallback.rs

```diff
@@ -0,0 +1,453 @@
+        }
+    }
+}
+
+impl<S> FallbackService<S>
+where
+    S: Service<RequestPacket, Future = TransportFut<'static>, Error = TransportError>
+        + Send
+        + Clone
+        + Debug
+        + 'static,
+{
+    /// Make a request to the fallback service middleware.
+    ///
+    /// Here is a high-level overview of how requests are handled:
+    ///
+    /// - At the start of each request, we assume `self.transports` to be sorted by score
+    /// - We take the top `self.active_transport_count` transports and call them in parallel
+    /// - If any of them succeeds, we re-sort `self.transports` to reflect the new scores
+    /// - We then return the first successful response and clear the remaining futures
+    /// - If all transports fail, we return the last error that occurred
+    ///
+    /// This strategy allows us to always make requests to the best available transports
+    /// while keeping them available. We also only update the transport scores after receiving
+    /// a response, avoiding unnecessary re-sorting.
+    async fn make_request(&self, req: RequestPacket) -> Result<ResponsePacket, TransportError> {
+        // Get the top transports to use, without removing them from the heap
+        let top_transports = {
+            let transports = self.transports.read();
+            transports.iter().take(self.active_transport_count).cloned().collect::<Vec<_>>()
+        };
+
+        // Create a collection of future requests
+        let mut futures = FuturesUnordered::new();
+
+        // Launch requests to all active transports in parallel
+        for transport in top_transports {
+            let start = Instant::now();
+            let req_clone = req.clone();
+            let mut transport_clone = transport.clone();
+
+            let future = async move {
+                let result = transport_clone.call(req_clone).await;
+                trace!(
+                    "Transport[{}] completed: latency={:?}, status={}",
+                    transport_clone.id,
+                    start.elapsed(),
+                    if result.is_ok() { "success" } else { "fail" }
+                );
+
+                (result, transport_clone, start.elapsed())
+            };
+
+            futures.push(future);
+        }
+
+        // Wait for the first successful response or until all fail
+        let mut last_error = None;
+
+        while let Some((result, transport, duration)) = futures.next().await {
+            match result {
+                Ok(response) => {
+                    // Record success
+                    {
+                        let mut metrics = transport.metrics.write();
+                        metrics.track_success(duration);
+                    }
+
+                    // Sort transports after updating metrics
+                    {
+                        let mut transports = self.transports.write();
+                        transports.sort_by(|a, b| b.cmp(a));
+                    }
```

Comment:
I think we can avoid this write here if we instead sort when we retrieve all the transports then we don't need an additional write lock here

like first clone all transports, sort them, then take only self.active_transport_count

Reply:
Got it, then I can drop the RwLock for the transports altogether!